src/platforms/Token/EditTokenModal.tsx

```diff
@@ -0,0 +1,318 @@
+
+  const circleBgColor = useColorModeValue("blackAlpha.200", "gray.600")
+  const [snapshotOption, setSnapshotOption] = useState(SnapshotOption.GUILD_POINTS)
+
+  const dynamicOptions: Option[] = [
+    {
+      value: SnapshotOption.GUILD_POINTS,
+      title: "Guild points snapshot",
+      description:
+        "Calculate rewards based on users' Guild points at a specific time",
+      leftComponent: (
+        <Circle bg={circleBgColor} p={3}>
+          <Icon as={Star} />
+        </Circle>
+      ),
+      children: <GuildPointsSnapshotForm />,
+    },
+    {
+      value: SnapshotOption.CUSTOM,
+      title: "Custom snapshot",
+      description:
+        "Upload a custom snapshot to assign unique numbers to users for reward calculation",
+      leftComponent: (
+        <Circle bg={circleBgColor} p={3}>
+          <Icon as={ListNumbers} />
+        </Circle>
+      ),
+      children: <CustomSnapshotForm />,
+    },
+  ]
+
+  const methods = useForm<AddTokenFormType>({
+    mode: "all",
+  })
+
+  const { setValue } = methods
+
+  const { roles, id: guildId } = useGuild()
+
+  const rolePlatforms = useRolePlatforms(id)
+
+  const role = roles.find((rl) =>
+    rl.rolePlatforms.find((rp) => rp.id === rolePlatforms[0].id)
+  )
+  const { data: requirements } = useRequirements(role.id)
+  const snapshotRequirement = requirements?.find((req) => !!req?.data?.snapshot)
+
+  useEffect(() => {
+    const rp: any = rolePlatforms[0]
+    const multiplier = rp.dynamicAmount.operation.params.multiplier || 1
+    setValue("multiplier", multiplier)
+  }, [rolePlatforms, setValue])
+
+  const {
+    isOpen: snapshotIsOpen,
+    onOpen: snapshotOnOpen,
+    onClose: snapshotOnClose,
+  } = useDisclosure()
+
+  const toast = useToast()
+  const showErrorToast = useShowErrorToast()
+
+  const { onSubmit: submitEditRolePlatform, isLoading: rpIsLoading } =
+    useEditRolePlatform({
+      rolePlatformId: rolePlatforms[0].id,
+      onSuccess: () => {
+        toast({
+          status: "success",
+          title: "Successfully updated the conversion rate!",
+        })
+      },
+    })
+
+  const { onSubmit: submitEditRequirement, isLoading: reqIsLoading } =
+    useEditRequirement(role.id, {
+      onSuccess: () => {
+        toast({
+          status: "success",
+          title: "Successfully changed the snapshot!",
+        })
+      },
+    })
+
+  const { onSubmit: onRequirementSubmit } = useCreateRequirementForRole()
+
+  const { mutateGuild } = useGuild()
+
+  const mutateRequirements = (
+    req: Requirement,
+    roleId: number,
+    idForGuild: number
+  ) => {
+    mutateOptionalAuthSWRKey<Requirement[]>(
+      `/v2/guilds/${idForGuild}/roles/${roleId}/requirements`,
+      (prevRequirements) => [
+        ...prevRequirements.filter((r) => r.type === "FREE"),
+        req,
+      ],
+      { revalidate: false }
+    )
+  }
+
+  const { triggerMembershipUpdate } = useMembershipUpdate()
+
+  const onEditSubmit = async (data) => {
```

Comment:
Could we type `data` here? I actually don't know it's format, so it would be good if TS could "help us" inside this function 😉